core/src/test/java/org/apache/iceberg/TestUpdatePartitionSpec.java

```diff
@@ -424,6 +424,38 @@ public void testAddDeletedName() {
     V2Assert.assertEquals("Should match expected spec", v2Expected, updated);
   }
 
+  @Test
+  public void testAddDeletedPartitionBack() {
+    PartitionSpec initialExpectedSpec = PartitionSpec.builderFor(table.schema())
+        .withSpecId(TableMetadata.INITIAL_SPEC_ID)
+        .bucket("data", 16)
+        .build();
+
+    Assert.assertEquals("Initial spec must match", initialExpectedSpec, table.spec());
+
+    // remove an existing partition column
+    table.updateSpec()
+        .removeField(bucket("data", 16))
+        .commit();
+
+    // add the same partition column back
+    table.updateSpec()
+        .addField("data_bucket", bucket("data", 16))
```

Comment:
**Attention**: I had to use the exact name to reuse the partition spec. It looks like the partition field name generation is different during the initial table creation and during spec evolution. 

For example, if I create a table partitioned by `bucket("data", 16)`, the name is `data_bucket`. If I add such a partition field, the assigned name will be `data_bucket_16`.

Reply:
This is correct. We changed how name generation was happening was to avoid conflicts between bucket partition names. For example, if you were moving from `bucket[16]` to `bucket[32]` and both used `data_bucket` then we would generate a conflicting name instead, evolution will produce `data_bucket_32`.